### ISSUE TITLE
depth issue in multimarker_pattern_threejs.html

### ISSUE THREAD

**Issue by jeromeetienne** [artoolkit/jsartoolkit5#2]
![screen shot 2016-02-09 at 06 28 44](https://cloud.githubusercontent.com/assets/252962/12909194/711a04a0-cef6-11e5-8de1-52c7e5445143.png)

in this examples, it seems to have a depth issue which make it flicker all the time.

**Comment by jeromeetienne** [comment]
hmm @philip-lamb im thinking it may be due to the camera parameter reading... what do you think ?

Supposition:
- like the camera generated from the parameter would have a near/far very distant from each other
- this would create a massive z-figthing

would that be possible ?

**Comment by philip-lamb** [comment]
Z-fighting would cause that, and yes, poorly chosen near/far would cause z-fighting. The camera parameter doesn't directly set the near/far planes, but the camera parameters are merged with the requested near/far values to create the projection matrix. I would need to investigate where those values are set.

**Comment by jeromeetienne** [comment]
@philip-lamb is there a way to read the near/far out of

**Comment by philip-lamb** [comment]
Near and far planes appear to be set here: https://github.com/artoolkit/jsartoolkit5/blob/master/js/artoolkit.api.js#L960

These need to be made configurable.

**Comment by jeromeetienne** [comment]
Changing those values doesnt affect the issue unfortunatly. This line seems to be the source of it

```
camera.projectionMatrix.elements.set(this.getCameraMatrix());
```

When i replace it with a approximation of it, all flickering issues disapear. This is bypassing camera parameters loading tho.

```
			camera = new THREE.PerspectiveCamera( 43, 640 / 480, 1, 2000 );
			camera.lookAt(new THREE.Vector3(0,0,1))
			camera.rotateZ(Math.PI)
```

**Comment by jeromeetienne** [comment]
my bet is that the near/far we use when we compute the projection matrix in jsartoolkit are too far apart.

im not sure where to check this supposition tho...

**Comment by yvt** [comment]
@jeromeetienne You are right; they default to `0.0001` and `1000.0`. And currently, there is no way to change them because `setProjectionNearPlane` and `setProjectionFarPlane` just store the new near/far values to member variables and do not actually recalculate the projection matrix.

**Comment by simonharrisco** [comment]
@jeromeetienne  I had a go at this yesterday, unfortunately the projection matrix calculations were hidden and unreadable by my skill as they seem to have been part of the emscripten translation. Regardless, I did manage to reverse engineer the projection matrix calculations:

`camera.projectionMatrix.elements.set(this.getCameraMatrix());`
 was getting its data from memory directly in this line:
`this.camera_mat = new Float64Array(Module.HEAPU8.buffer, params.camera, 16);`

I overwrote this with the projection calculation I wrote and ended up with something somewhat messy but usable

`	//this.camera_mat = new Float64Array(Module.HEAPU8.buffer, params.camera, 16);

		 this.projectionMatrixCalc = (fieldOfViewYInRadians, aspect, zNear, zFar, dst)  => {
		    dst = dst || new Float64Array(16);

		    var f = Math.tan(Math.PI * 0.5 - 0.5 * fieldOfViewYInRadians);
		    var rangeInv = 1.0 / (zNear - zFar);

		    dst[0]  = f / aspect;
		    dst[1]  = 0;
		    dst[2]  = 0;
		    dst[3]  = 0;

		    dst[4]  = 0;
		    dst[5]  = -f;
		    dst[6]  = 0;
		    dst[7]  = 0;

		    dst[8]  = 0;
		    dst[9]  = 0;
		    dst[10] = -(zNear + zFar) * rangeInv;
		    dst[11] = 1;

		    dst[12] = 0;
		    dst[13] = 0;
		    dst[14] = zNear * zFar * rangeInv * 2;
		    dst[15] = 0;

		    return dst;


		  }
		this.camera_mat = this.projectionMatrixCalc(0.750492, 640 / 480, 0.1, 100,);`

As you can see the near and far clipping planes can be set here so this could perhaps be integrated somewhere a bit neater

**Comment by mciastek** [comment]
This problem is still relevant. It's not only an issue with multiple markers, but also with overlapping geometries. I noticed it, when I loaded model with overlapping outline mesh.

I used solution provided by @simonharrisco, but I'm not sure if it's accurate for all device cameras. Is it possible to overwrite `near` and `far` of camera matrix provided by ARToolkit?
